pysite/views/wiki/edit.py

```diff
@@ -48,10 +73,104 @@ def post(self, page):
             "headers": rendered["headers"]
         }
 
+        self.audit_log(page, obj)
+
         self.db.insert(
             self.table_name,
             obj,
             conflict="replace"
         )
 
+        # Add the post to the revisions table
+        revision_payload = {
+            "slug": page,
+            "post": obj,
+            "date": datetime.datetime.utcnow().timestamp(),
+            "user": self.user_data.get("user_id")
+        }
+
+        del revision_payload["post"]["slug"]
+
+        self.db.insert(self.revision_table_name, revision_payload)
+
         return redirect(url_for("wiki.page", page=page), code=303)  # Redirect, ensuring a GET
+
+    @require_roles(*EDITOR_ROLES)
+    @csrf
+    def patch(self, page):
+        current = self.db.get(self.table_name, page)
+        if not current:
+            return "", 404
+
+        if current.get("lock_expiry"):  # If there is a lock present
+
+            # If user patching is not the user with the lock end here
+            if current["lock_user"] != self.user_data.get("user_id"):
+                return "", 400
+            new_lock = datetime.datetime.utcnow() + datetime.timedelta(minutes=5)  # New lock time, 5 minutes in future
+            self.db.insert(self.table_name, {
+                "slug": page,
+                "lock_expiry": new_lock.timestamp()
+            }, conflict="update")  # Update with new lock time
+        return "", 204
+
+    def audit_log(self, page, obj):
+        if WIKI_AUDIT_WEBHOOK:  # If the audit webhook is not configured there is no point processing the diff
+            before = self.db.get(self.table_name, page)
+            if not before:  # If this is a new page, before will be None
+                before = []
+            else:
+                if before.get("rst") is None:
+                    before = []
+                else:
+                    before = before["rst"].splitlines(keepends=True)
+                    if len(before) == 0:
+                        pass
+                    else:
```

Comment:
Can this be an `elif`, since there's only one if clause within this `else`?

Reply:
```py
if not before[-1].endswith("\n"):
    before[-1] += "\n"  # difflib sometimes messes up if a newline is missing on last line
```
This line will error if the len is zero hence why I check for the length of the list